charts/argo-workflows/templates/controller/workflow-controller-cluster-roles.yaml

```diff
@@ -154,7 +154,12 @@ rules:
   - update
   - patch
   - delete
-
+- apiGroups:
```

Comment:
Couple of notes here:
1. The HTTP template requests are executed in [Argo Agent](https://argoproj.github.io/argo-workflows/http-template/#argo-agent), which is a special pod created alongside other workflow pods. It is being run with service account name that is [different](https://github.com/argoproj/argo-workflows/blob/d33d391a4c06c136b6a0964a51c75850323684e6/workflow/controller/agent.go#L151) from the one the controller runs under. Thus, grunting this permission to the controller makes no sense.
2. Even when applied to the workflow role, this file generates a permission to read all secrets in the namespace. This is quite insecure, and such a chart will be unusable in many organizations that care about security. Instead, the role should implement support for a secret whitelist, similar to how it's [done](https://github.com/argoproj/argo-helm/blob/4ed76386e00d636eb906eef86ab77bdf54ea0915/charts/argo-workflows/templates/server/server-cluster-roles.yaml#L71) in the server, but it should be secure by default.

Because

Reply:
Thank you for teaching me the specification and reviewing ! 🙇 

So, at least I need to handle below.
- grunt this permission to the right service account
- the target of reading secrets should be limited as whitelist, instead of reading all secrets.

I will fix them. 🙋

Reply:
According to logs below and logic to get secret ( [here](https://github.com/argoproj/argo-workflows/blob/d33d391a4c06c136b6a0964a51c75850323684e6/workflow/controller/agent.go#L74-L83) ), it looks that workflow controller gets secret, not argo agent.
So I think we need to implement support for a secret whitelist on workflow-controller-cluster-roles...? 🤔 

```
# when submitting http templates workflowTemplate

time="2022-06-18T02:50:23.799Z" level=error msg="error in agent pod reconciliation" error="failed to check if secret argo-workflows-agent-ca-certificates exists: secrets \"argo-workflows-agent-ca-certificates\" is forbidden: User \"system:serviceaccount:default:argo-argo-workflows-workflow-controller\" cannot get resource \"secrets\" in API group \"\" in the namespace \"default\"" namespace=default workflow=http-template-8ddc7-cljhd
time="2022-06-18T02:50:23.802Z" level=info msg="Updated phase Running -> Error" namespace=default workflow=http-template-8ddc7-cljhd
time="2022-06-18T02:50:23.802Z" level=info msg="Updated message  -> failed to check if secret argo-workflows-agent-ca-certificates exists: secrets \"argo-workflows-agent-ca-certificates\" is forbidden: User \"system:serviceaccount:default:argo-argo-workflows-workflow-controller\" cannot get resource \"secrets\" in API group \"\" in the namespace \"default\"" namespace=default workflow=http-template-8ddc7-cljhd
time="2022-06-18T02:50:23.803Z" level=info msg="Marking workflow completed" namespace=default workflow=http-template-8ddc7-cljhd
time="2022-06-18T02:50:23.809Z" level=info msg="cleaning up pod" action=deletePod key=default/http-template-8ddc7-cljhd-1340600742-agent/deletePod
time="2022-06-18T02:50:23.803Z" level=info msg="Checking daemoned children of " namespace=default workflow=http-template-8ddc7-cljhd
time="2022-06-18T02:50:23.892Z" level=info msg="Update workflows 200"
time="2022-06-18T02:50:23.896Z" level=info msg="Create events 201"
time="2022-06-18T02:50:23.870Z" level=info msg="Delete pods 404"
time="2022-06-18T02:50:23.931Z" level=info msg="Workflow update successful" namespace=default phase=Error resourceVersion=2832 workflow=http-template-8ddc7-cljhd
```

Reply:
@vladlosev What do you think about above? 👀

Reply:
I stand corrected. Thanks for straightening me out! It looks like the controller also needs access to a secret when HTTP templates are involved. But apparently, the name of that secret is [hardcoded](https://github.com/argoproj/argo-workflows/blob/d33d391a4c06c136b6a0964a51c75850323684e6/workflow/controller/agent.go#L126), so you need to grant the controller the right to read only that secret.

Reply:
Thank you for checking. I added read policy for hardcoded secret (`argo-workflows-agent-ca-certificates` ). 🙋

Reply:
This will unblock the workflow controller, but the agent pod needs the permissions, as well. Do you want to do them in a separate PR?

Reply:
`argo-workflows-agent-ca-certificates` secret is fetched by only workflow controller and agent uses the fetched secret, so I thought agent pod did not need the permission... I misunderstood it.
I will continue on this PR. 🙋

Reply:
> the agent pod needs the permissions, as well

Though I could not test it on my local environment for now because of below, adding this permission to [argo-workflow](https://github.com/argoproj/argo-helm/blob/main/charts/argo-workflows/templates/controller/workflow-role.yaml) role would be ok... ? 🤔 

```yaml
- apiGroups:
  - ""
  resources:
  - secrets
  verbs:
  - get
  - watch
  - list
  resourceNames:
  - argo-workflows-agent-ca-certificates
```

---

### What I tried to test
I created `argo-workflow` ServiceAccount for workflow service account, by enabling `.Values.workflow.serviceAccount.create`. Then I added `read secret` policy to `argo-workflow` sa. 
The role that is tied to `argo-workflow` sa is below.

```
$ k describe role argo-argo-workflows-workflow
Name:         argo-argo-workflows-workflow
Labels:       app.kubernetes.io/managed-by=Helm
Annotations:  meta.helm.sh/release-name: argo
              meta.helm.sh/release-namespace: argo
PolicyRule:
  Resources                            Non-Resource URLs  Resource Names                          Verbs
  ---------                            -----------------  --------------                          -----
  pods/exec                            []                 []                                      [create]
  workflowtaskresults.argoproj.io      []                 []                                      [create]
  secrets                              []                 [argo-workflows-agent-ca-certificates]  [get watch list]
  pods                                 []                 []                                      [get watch patch]
  pods/log                             []                 []                                      [get watch]
  workflowtasksets.argoproj.io         []                 []                                      [list watch patch]
  workflowtasksets.argoproj.io/status  []                 []                                      [patch]
```

I submitted workflow below and failed with `failed to get token volumes: service account default/argo-workflow does not have any secrets` . Actually I could not find any secrets tied to sa. Maybe because I tested on [kind](https://kind.sigs.k8s.io/) cluster..? 🤔 

```yaml
# This template is from https://argoproj.github.io/argo-workflows/http-template/
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: http-template-
spec:
  serviceAccountName: argo-workflow # created by `.Values.workflow.serviceAccount.create`
  entrypoint: main
  templates:
    - name: main
      steps:
        - - name: get-google-homepage
            template: http
            arguments:
              parameters: [{name: url, value: "https://www.google.com"}]
    - name: http
      inputs:
        parameters:
          - name: url
      http:
        timeoutSeconds: 20 # Default 30
        url: "{{inputs.parameters.url}}"
        method: "GET" # Default GET
        headers:
          - name: "x-header-name"
            value: "test-value"
        successCondition: "response.body contains \"google\"" # available since v3.3
        body: "test body" # Change request body
```

Errors when I submitted workflow. This errors occurred [here](https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/agent.go#L152).

```
time="2022-06-21T14:33:18.791Z" level=info msg=reconcileAgentPod namespace=default workflow=http-template-fq8t2
time="2022-06-21T14:33:18.792Z" level=info msg="Get secrets 200"
time="2022-06-21T14:33:18.793Z" level=info msg="Get serviceaccounts 200"
time="2022-06-21T14:33:18.793Z" level=error msg="error in agent pod reconciliation" error="failed to get token volumes: service account default/argo-workflow does not have any secrets" namespace=default workflow=http-template-fq8t2
time="2022-06-21T14:33:18.793Z" level=info msg="Updated phase Running -> Error" namespace=default workflow=http-template-fq8t2
time="2022-06-21T14:33:18.793Z" level=info msg="Updated message  -> failed to get token volumes: service account default/argo-workflow does not have any secrets" namespace=default workflow=http-template-fq8t2
time="2022-06-21T14:33:18.793Z" level=info msg="Marking workflow completed" namespace=default workflow=http-template-fq8t2
time="2022-06-21T14:33:18.793Z" level=info msg="Checking daemoned children of " namespace=default workflow=http-template-fq8t2
time="2022-06-21T14:33:18.793Z" level=debug msg="Log changes patch: {\"metadata\":{\"annotations\":{\"workflows.argoproj.io/pod-name-format\":\"v1\"},\"labels\":{\"workflows.argoproj.io/completed\":\"true\",\"workflows.argoproj.io/phase\":\"Error\"}},\"status\":{\"artifactRepositoryRef\":{\"artifactRepository\":{},\"default\":true},\"conditions\":[{\"status\":\"True\",\"type\":\"Completed\"}],\"finishedAt\":\"2022-06-21T14:33:18Z\",\"message\":\"failed to get token volumes: service account default/argo-workflow does not have any secrets\",\"nodes\":{\"http-template-fq8t2\":{\"children\":[\"http-template-fq8t2-1133102788\"],\"displayName\":\"http-template-fq8t2\",\"finishedAt\":null,\"id\":\"http-template-fq8t2\",\"message\":\"Workflow operation error\",\"name\":\"http-template-fq8t2\",\"phase\":\"Error\",\"startedAt\":\"2022-06-21T14:33:18Z\",\"templateName\":\"main\",\"templateScope\":\"local/http-template-fq8t2\",\"type\":\"Steps\"},\"http-template-fq8t2-1133102788\":{\"boundaryID\":\"http-template-fq8t2\",\"children\":[\"http-template-fq8t2-584265465\"],\"displayName\":\"[0]\",\"finishedAt\":null,\"id\":\"http-template-fq8t2-1133102788\",\"name\":\"http-template-fq8t2[0]\",\"phase\":\"Running\",\"startedAt\":\"2022-06-21T14:33:18Z\",\"templateScope\":\"local/http-template-fq8t2\",\"type\":\"StepGroup\"},\"http-template-fq8t2-584265465\":{\"boundaryID\":\"http-template-fq8t2\",\"displayName\":\"get-google-homepage\",\"finishedAt\":null,\"id\":\"http-template-fq8t2-584265465\",\"inputs\":{\"parameters\":[{\"name\":\"url\",\"value\":\"https://www.google.com\"}]},\"name\":\"http-template-fq8t2[0].get-google-homepage\",\"phase\":\"Pending\",\"startedAt\":\"2022-06-21T14:33:18Z\",\"templateName\":\"http\",\"templateScope\":\"local/http-template-fq8t2\",\"type\":\"HTTP\"}},\"phase\":\"Error\",\"startedAt\":\"2022-06-21T14:33:18Z\"}}"
time="2022-06-21T14:33:18.794Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"default\", Name:\"http-template-fq8t2\", UID:\"9632d55a-46be-4661-9b1b-aa91afb694c6\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"8347\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowFailed' failed to get token volumes: service account default/argo-workflow does not have any secrets"
time="2022-06-21T14:33:18.796Z" level=info msg="Create events 201"
time="2022-06-21T14:33:18.797Z" level=info msg="Update workflows 200"
time="2022-06-21T14:33:18.798Z" level=info msg="Workflow update successful" namespace=default phase=Error resourceVersion=8351 workflow=http-template-fq8t2
time="2022-06-21T14:33:18.798Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"default\", Name:\"http-template-fq8t2\", UID:\"9632d55a-46be-4661-9b1b-aa91afb694c6\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"8351\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeRunning' Running node http-template-fq8t2: Workflow operation error"
time="2022-06-21T14:33:18.798Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"default\", Name:\"http-template-fq8t2\", UID:\"9632d55a-46be-4661-9b1b-aa91afb694c6\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"8351\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeError' Error node http-template-fq8t2: Workflow operation error"
time="2022-06-21T14:33:18.798Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"default\", Name:\"http-template-fq8t2\", UID:\"9632d55a-46be-4661-9b1b-aa91afb694c6\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"8351\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeRunning' Running node http-template-fq8t2[0]"
```

Reply:
Hi @vladlosev ,
After I added `get secret` policy to workflow controller, I tested on minikube and it worked well without any additional policy to argo agent (Service Account).
Did I miss or misunderstand something ..? 🤔 

<img width="1910" alt="Screen Shot 2022-06-22 at 20 28 17" src="https://user-images.githubusercontent.com/30188755/175018430-a8f1afb2-47d5-47aa-8d5c-c4d13584d6e6.png">

- values.yaml on my customized Helm Chart
```yaml
argo-workflows:
  controller:
    logging:
      level: debug
  server:
    secure: false
  workflow:
    serviceAccount:
      create: true # `argo-workflow` Service Account is created and it is for attaching workflow
```

- workflow template (same as [here](https://argoproj.github.io/argo-workflows/http-template/), except for `serviceAccountName` )
```yaml
spec:
  templates:
    - name: main
      steps:
        - - name: get-google-homepage
            template: http
            arguments:
              parameters:
                - name: url
                  value: 'https://www.google.com'
    - name: http
      inputs:
        parameters:
          - name: url
      http:
        method: GET
        url: '{{inputs.parameters.url}}'
        headers:
          - name: x-header-name
            value: test-value
        timeoutSeconds: 20
        successCondition: response.body contains "google"
        body: test body
  entrypoint: main
  arguments: {}
  serviceAccountName: argo-workflow
```

- Service Account situation
  - It does not have any policy for secrets
```
$ k describe rolebinding argo-argo-workflows-workflow
Name:         argo-argo-workflows-workflow
Labels:       app.kubernetes.io/managed-by=Helm
Annotations:  meta.helm.sh/release-name: argo
              meta.helm.sh/release-namespace: argo
Role:
  Kind:  Role
  Name:  argo-argo-workflows-workflow
Subjects:
  Kind            Name           Namespace
  ----            ----           ---------
  ServiceAccount  argo-workflow  default

$ k describe role argo-argo-workflows-workflow
Name:         argo-argo-workflows-workflow
Labels:       app.kubernetes.io/managed-by=Helm
Annotations:  meta.helm.sh/release-name: argo
              meta.helm.sh/release-namespace: argo
PolicyRule:
  Resources                            Non-Resource URLs  Resource Names  Verbs
  ---------                            -----------------  --------------  -----
  pods/exec                            []                 []              [create]
  workflowtaskresults.argoproj.io      []                 []              [create]
  pods                                 []                 []              [get watch patch]
  pods/log                             []                 []              [get watch]
  workflowtasksets.argoproj.io         []                 []              [list watch patch]
  workflowtasksets.argoproj.io/status  []                 []              [patch]
```